core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushJoinIntoTableScan.java

```diff
@@ -162,6 +169,43 @@ public Result apply(JoinNode joinNode, Captures captures, Context context)
         return Result.ofPlanNode(new TableScanNode(joinNode.getId(), handle, joinNode.getOutputSymbols(), newAssignments.build(), newEnforcedConstraint, false));
     }
 
+    private boolean skipJoinPushdownBasedOnCost(JoinNode joinNode, Context context)
+    {
+        if (getJoinPushdownMode(context.getSession()) != JoinPushdownMode.AUTOMATIC) {
+            return false;
+        }
+
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+
+        // returning as quickly as possible to avoid unnecessary, costly work
+
+        PlanNodeStatsEstimate leftStats = context.getStatsProvider().getStats(joinNode.getLeft());
+        double leftOutputSize = leftStats.getOutputSizeInBytes(joinNode.getLeft().getOutputSymbols(), types);
+        if (isNaN(leftOutputSize)) {
+            return true;
+        }
+
+        PlanNodeStatsEstimate rightStats = context.getStatsProvider().getStats(joinNode.getRight());
+        double rightOutputSize = rightStats.getOutputSizeInBytes(joinNode.getRight().getOutputSymbols(), types);
+        if (isNaN(rightOutputSize)) {
+            return true;
+        }
+
+        PlanNodeStatsEstimate joinStats = context.getStatsProvider().getStats(joinNode);
+        double joinOutputSize = joinStats.getOutputSizeInBytes(joinNode.getOutputSymbols(), types);
+        if (isNaN(joinOutputSize)) {
+            return true;
+        }
+
+        if (joinOutputSize > leftOutputSize + rightOutputSize) {
```

Comment:
Consider adding some factor here, e.g pushed down join should produce 2x less rows than in trino. Such factor might need to be empirically established

Reply:
so you mean to replace `left + right` with `max(left, right) * 0.5`?  Works for me, given that the current formula is not very scientificly determined.
I think we should do "something reasonable" & iterate.

Reply:
Yeah - I find initial value of a factor `1.0` as good as `0.5`